rfcs/protocols/asset-exchange/corda-htlc-alternatives.md

```diff
@@ -0,0 +1,68 @@
+
+#### Hash Lock Encumbrance State
+A generalised state class that encodes the properties of the HTLC such as the image of the secret, the time window within which a counter party can redeem the asset, and the public key of all involved parties as well as a reference to the binding contract logic. This state will be attached to an asset as an encumbrance, to conditionally lock the underlying asset.
+<img src="../../resources/images/asset-exchange-corda-encumbrance-lock.png" width=80%>
+
+#### Hash Lock Encumbrance Validation Contract
+The validation contract ensures that all attempts to spend the underlying asset is validated against the conditions of the hashed lock encumbrance state. That is, it ensures that only the owner or the designated counter party can make claims against the event, that the counter-party can only claim the state upon presentation of the pre-image of the stored secret within the specified validity window, and that the owner can only claim a refund of the asset if the specified time period has elapsed.
+
+#### Process
+The procedures for setting up an HTLC with one Corda network, which would need to be mirrored in the counter party network, involves the following:
+- Party A and B agree on terms of the contract (i.e. secret, time window, asset amount etc.)
+- Party A creates a transaction that takes its asset state as input and produces i) a hash lock encumbrance state and ii) the asset state with a reference to the encumbrance. Party A sends this transaction to Party B.
+- **Claiming**: Party B submits a transaction to claim the locked asset with the pre-image of the secret. If the pre-image is valid and the transaction is within the defined time window (as validated by the encumbrance contract), the output of the transaction will be a new state representing the asset with Party B as its owner that no longer has the encumbrance state attached to it.
+<img src="../../resources/images/asset-exchange-corda-encumbrance-claim.png" width=80%>
+- **Refund**: Party A submits a transaction to claim back its locked asset. If this claim occurs after the defined window has elapsed, then a new state representing the state with Part A as the owner is output that no longer has the encumbrance state attached to it.
+
+Both claiming and refunding transactions can be performed unilaterally because during the locking phase the asset is associated with a Composite key of the two parties.
+
+![Corda HTLC](../../resources/images/cordapp-htlc.png)
+
+### Generalisability
+The hash lock encumbrance state and associated contract can be offered as reusable pieces as part of Weaver. These pieces are largely agnostic of the specifics of the application. They can further be instrumented to *emit* custom Weaver events making automation of the orchestration of cross-chain HTLCs possible.
```

Comment:
Can you add another section following this where you compare the pros and cons of the two mechanisms (validating notary vs encumbrance states), state that we've picked the former for our current implementation in Weaver (link to the interop CorDapp source code folder), and summarize why we picked that (encumbrances not being safe w.r.t. contract upgrades, etc.)?